content/2019/Making-things-flow/index.md

```diff
@@ -0,0 +1,98 @@
+
+```vim
+" When editing a file, always jump to the last known cursor position.
+" Don't do it when the position is invalid, when inside an event handler
+" (happens when dropping a file on gvim) and for a commit message (it's
+" likely a different one than last time).
+autocmd BufReadPost *
+  \ if line("'\"") >= 1 && line("'\"") <= line("$") && &ft !~# 'commit'
+  \ |   exe "normal! g`\""
+  \ | end
+```
+
+The comment explains the what's being done in the `autocmd` so I won't talk about that. But it's interesting to think about the fact that this is something Vim now ships with. It's been possible and indeed used by others in their configuration before appearing in mainline Vim. I consider this interesting for two reasons, firstly Vim scripting allows us to do this quite easily and secondly the scripting was added to Vim rather than code for an option that would allow users to `set` such behaviour.
+
+## Getting into the groove
+
+Whilst having been a Vim user for some time my first real exposure to the idea of curving Vim's behaviour in a transparent way was this excellent [gist][CCR]. I thoroughly recommend reading through it then coming back here. I actually even made my own [variant][Autoreply].
+
+In any case the effect is the same. Vim's behaviour is altered not because of a setting but because we've scripted a flow. It's something we can use a great deal but we've only ever had to think about it once, when we wrote the script. To contrast we don't have to think "Oh I have a mapping that will show me X and prompt for Y", instead we ask Vim for X as usual it now prompts for Y.
+
+Continuing with a another "call and response" type example we can have the quickfix window open when it contains results.
+
+Here's a [gist][instant-grep] that showcases this. This is also behaviour is also available in [vim-qf][vim-qf].
+
+This method is very useful as there are many ways the quickfix list may be populated and yet we are able to create a behaviour that is consistent with a self contained implementation.
+
+## Don't disrupt my flow
+
+Disabling `hlsearch` is a common operation for users that have `set hlsearch` as after performing the search and finding the desired result the highlighting becomes superfluous. To ease this many Vim users have a mapping to deal with this. But that's an extra step that the user has to think about. Instead we can alter the behaviour of Vim to intelligently disable and re-enable `hlsearch` as done with [vim-cool][vim-cool].
+
+As noted in the README this plugin really became "cool" with [this][vim-cool_purpleP]. It doesn't change the intention of the plugin but it achieves it in a far robust and elegant way.
+
+## Don't break my flow
+
+Another example is the movement of the cursor when dealing with operators. Vim has moves the cursor after invoking operators and I'd prefer it didn't always do so. Here we wish to prevent a behaviour default to Vim rather than craft a complementary one.
```

Comment:
```suggestion
Another example is the movement of the cursor when dealing with operators. Vim moves the cursor after invoking operators and I'd prefer it didn't always do so. Here we wish to prevent a behaviour default to Vim rather than craft a complementary one.
```